### PR TITLE
Implement core algorihtm

### PR DESCRIPTION
**Pull request by tomaslink**

### CODE REVIEW

File: pipe_gaps/core/gap_detector.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    logger.info("Sorting messages by timestamp...")
+    messages_sorted = sorted(messages, key=lambda x: x["timestamp"])
+
+    logger.info("Detecting gaps...")
+    gaps = zip(messages_sorted[:-1], messages_sorted[1:])
+    gaps = tuple(
+        gap for gap in track(gaps, total=n, description="Filtering gaps:")
+        if _filter_condition(gap, threshold)
+    )
```
Comment (andres-arana) [inline pipe_gaps/core/gap_detector.py:37]: Consider using itertools for this kind of iteration.

File: pipe_gaps/core/gap_detector.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    logger.info("Amount of gaps found: {}".format(len(gaps)))
+
+    return gaps
+
+
+def _filter_condition(gap: tuple[dict, dict], threshold: timedelta) -> bool:
+    start_dt = ciso8601.parse_datetime(gap[0]["timestamp"].replace(" UTC", ""))
+    end_dt = ciso8601.parse_datetime(gap[1]["timestamp"].replace(" UTC", ""))
+
+    return (end_dt - start_dt) > threshold
```
Comment (andres-arana) [inline pipe_gaps/core/gap_detector.py:48]: Sometimes what we do to skip this is to convert timestamps to actual timestamps in the BQ query, and then you can use the dt in your core logic.

File: pipe_gaps/core/gap_detector.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    n = len(messages)
+    logger.info("Amount of messages: {}".format(n))
+
+    logger.info("Sorting messages by timestamp...")
+    messages_sorted = sorted(messages, key=lambda x: x["timestamp"])
+
+    logger.info("Detecting gaps...")
+    gaps = zip(messages_sorted[:-1], messages_sorted[1:])
+    gaps = tuple(
+        gap for gap in track(gaps, total=n, description="Filtering gaps:")
```
Comment (andres-arana) [inline pipe_gaps/core/gap_detector.py:35]: You might want to disable this via a parameter (the progress bar I mean).

**Review by andres-arana — APPROVED** [review]
I left some small improvements you can implement, but overall LGTM

File: pipe_gaps/core/gap_detector.py (hunk 2 above)
Comment (tomaslink) [reply pipe_gaps/core/gap_detector.py]: Done! Now is almost 2x faster.

File: pipe_gaps/core/gap_detector.py (hunk 3 above)
Comment (tomaslink) [reply pipe_gaps/core/gap_detector.py]: Done

**Pull request merged**